src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs

```diff
@@ -3,11 +3,175 @@
+                                        last = castCurrent;
+                                    }
+                                }
+
+                                return last;
+                            }
+                        }
+                        while (e.MoveNext());
+                    }
+                }
+                finally
+                {
+                    (e as IDisposable)?.Dispose();
+                }
+
+                found = false;
+                return default;
+            }
+
+            public override TResult? TryGetElementAt(int index, out bool found)
+            {
+                if (index >= 0)
+                {
+                    foreach (object? item in _source)
+                    {
+                        if (item is TResult castItem)
+                        {
+                            if (index == 0)
+                            {
+                                found = true;
+                                return castItem;
+                            }
+
+                            index--;
+                        }
+                    }
+                }
+
+                found = false;
+                return default;
+            }
+
+            public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> selector)
+            {
+                // If the source is any generic enumerable of a reference type, which should be the 90% case, it'll covariantly
+                // implement IEnumerable<object>, and we can optimize the OfType().Select case by treating the OfType instead like
+                // a Where, using the same WhereSelectIterators that are used for Where.Select.
+                if (!typeof(TResult).IsValueType && _source is IEnumerable<object> objectSource)
+                {
+                    // Unsafe.As here is safe because we're only dealing with reference types, and we know by construction that only
+                    // TResult instances will be passed in. Using Unsafe.As allows us to avoid an extra closure and delegate allocation.
+                    Func<object, TResult2> localSelector =
+#if DEBUG
+                        o =>
+                        {
+                            Debug.Assert(o is TResult);
+                            return selector((TResult)o);
+                        };
+#else
+                        Unsafe.As<Func<object, TResult2>>(selector);
+#endif
+
+                    // We can special-case arrays and IEnumerable to use the corresponding WhereSelectIterators because
+                    // they're covariant. It's not worthwhile checking for List<T> to use the ListWhereSelectIterator
+                    // because List<> is not covariant.
+                    Func<object, bool> isTResult = static o => o is TResult;
+                    return objectSource is object[] array ?
+                        new ArrayWhereSelectIterator<object, TResult2>(array, isTResult, localSelector) :
```

Comment:
Could allocating the intermediate iterator be avoid somehow? E.g. by making the particular `Select` implementations static if possible?

Reply:
Which intermediate are you referring to? At this point we already have the OfType iterator, and now that someone is calling Select, we have to return a new object; mutating the current object would be visible and dangerous.

Reply:
I misunderstood what this method is doing, nevermind.